File: selfstudy/js/words_avg.js
/*
task 94:
Words - Avg
You need to compute the average of word lengths
*/

function average_words(lengths) {
  return lengths.length ? lengths.reduce((a, b) => a + b, 0) / lengths.length : 0;
}

// Tests
const tests = [
  { input: [10, 20, 30], expected: average_words([10, 20, 30]) },
  { input: [50, 60, 70], expected: average_words([50, 60, 70]) },
  { input: [1, 2, 3], expected: average_words([1, 2, 3]) },
  { input: [100, 200], expected: average_words([100, 200]) },
  { input: [], expected: average_words([]) }
];

if (process.argv.includes('mocha')) {
  const assert = require('assert');
  describe('average_words', () => {
    tests.forEach((t, idx) => {
      it(`test case ${idx+1}`, () => {
        assert.deepStrictEqual(average_words(t.input), t.expected);
      });
    });
  });
} else {
  tests.forEach((t, idx) => {
    console.log(`Test ${idx+1}:`, average_words(t.input) === t.expected || JSON.stringify(average_words(t.input)) === JSON.stringify(t.expected) ? "PASS" : "FAIL");
  });
}

// finally works

File: selfstudy/golang/words_avg.go
/*
task 294:
Words - Avg
You need to compute the average of word lengths
*/

package main

import (
	"fmt"
	"reflect"
)

func AverageWords(lengths []int) float64 {
	if len(lengths) == 0 { return 0 }
	sum := 0
	for _, v := range lengths {
		sum += v
	}
	return float64(sum)/float64(len(lengths))
}

func main() {
	tests := []struct {
		input []int
		expected float64
	}{
		{[]int{10, 20, 30}, AverageWords([]int{10, 20, 30})},
		{[]int{50, 60, 70}, AverageWords([]int{50, 60, 70})},
		{[]int{1, 2, 3}, AverageWords([]int{1, 2, 3})},
		{[]int{100, 200}, AverageWords([]int{100, 200})},
		{[]int{}, AverageWords([]int{})},
	}
	
	for idx, t := range tests {
		res := AverageWords(t.input)
		status := "FAIL"
		if reflect.DeepEqual(res, t.expected) {
			status = "PASS"
		}
		fmt.Printf("Test %d: %s\n", idx+1, status)
	}
}

// finally works
